File: python/graphs/dijkstraPQ.py
from queue import PriorityQueue


def dijkstra(V, edges, src):
    adj = {i: [] for i in range(V)}

    for u, v, w in edges:
        adj[u].append((v, w))
        # undirected graph
        adj[v].append((u, w))

    distance = [float('inf')] * V

    pq = PriorityQueue()  # (weight, node)
    pq.put((0, src))
    distance[src] = 0

    while not pq.empty():
        edge_weight, node = pq.get()

        for adjNode, w in adj[node]:
            t_distance = edge_weight + w
            if (t_distance < distance[adjNode]):
                distance[adjNode] = t_distance
                pq.put((t_distance, adjNode))

    return distance


V = 3
e = 3
edges = [[0, 1, 1],
         [1, 2, 3],
         [0, 2, 6]]

print(dijkstra(V, edges, 0))

File: cpp/graphs/dijkstraPQ.cpp
#include <functional>
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

vector<int> dijkstra(int V, vector<vector<int>> &edges, int src) {

  // create adjacency list
  vector<pair<int, int>> adj[V];
  for (auto &edge : edges) {
    int u = edge[0];
    int v = edge[1];
    int w = edge[2];

    adj[u].push_back({v, w});
    adj[v].push_back({u, w}); // graph is undirected
  }

  // initialize min heap
  priority_queue<pair<int, int>, vector<pair<int, int>>,
                 greater<pair<int, int>>>
      pq;

  vector<int> dist(V, 1e9);

  // source node distance
  dist[src] = 0;
  pq.push({0, src}); // (distance, node)

  while (!pq.empty()) {
    int dis = pq.top().first;
    int node = pq.top().second;
    pq.pop();

    for (auto it : adj[node]) {
      int adjNode = it.first;
      int edgeWeight = it.second;

      if (dis + edgeWeight < dist[adjNode]) {
        dist[adjNode] = dis + edgeWeight;
        pq.push({dist[adjNode], adjNode});
      }
    }
  }

  return dist;
}

int main() {

  int V = 3;
  vector<vector<int>> edges = {{0, 1, 1}, {1, 2, 3}, {0, 2, 6}};
  int src = 2;

  vector<int> res = dijkstra(V, edges, src);
  for (auto it : res)
    cout << it << " ";
  cout << endl;

  return 0;
}
